### PR TITLE
adapter new metric framework

### PR DESCRIPTION
**Pull request by MouceL**
adapter new metric framwork

### CODE REVIEW

File: controllers/process.go [hunk 1]
```diff
@@ -249,6 +308,7 @@ func refreshEnvoyFilter(instance *microservicev1alpha2.SmartLimiter, r *SmartLim
 // if configmap rate-limit-config not exist, ratelimit server will not running
 func refreshConfigMap(desc []*model.Descriptor, r *SmartLimiterReconciler, serviceLoc types.NamespacedName) {
 
+	// TODO, add switch
```
Comment (cywang1905) [inline controllers/process.go:311]: what's this todo mean, explain more clearly

File: controllers/producer.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+// StaticMeta is static info and do not to query from prometheus
+type StaticMeta struct {
+	Name string  	`json:"name"`
+	Namespace string `json:"namespace"`
+	NPod map[string]int `json:"nPod"`
+	IsGroup map[string]bool `json:"isGroup"`
+}
+
+func (n StaticMeta) String() string {
```
Comment (cywang1905) [inline controllers/producer.go:33]: sm instead of n is better?

File: controllers/producer.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+// some metric is not query from prometheus, so add it to staticMeta
+func generateMeta(subsetsPods map[string][]string, loc types.NamespacedName) StaticMeta {
+	// NPOD record like
+	// _base.pod: 6
+	// v1.pod: 2
+	// v2.pod: 4
+
+	NPod := make(map[string]int)
```
Comment (cywang1905) [inline controllers/producer.go:205]: nPod for camel style?

File: controllers/smartlimiter_controller.go [hunk 4]
```diff
@@ -108,7 +113,11 @@ func (r *SmartLimiterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 			r.lastUpdatePolicyLock.Lock()
 			r.lastUpdatePolicy = instance.Spec
 			r.lastUpdatePolicyLock.Unlock()
-			r.source.WatchAdd(req.NamespacedName)
+			r.interest.Set(req.Namespace+"/"+req.Name, struct {}{})
+			//r.handleEvent(types.NamespacedName{
+			//	Namespace: req.Namespace,
+			//	Name:      req.Name,
+			//})
```
Comment (cywang1905) [inline controllers/smartlimiter_controller.go:120]: remove if useless

File: controllers/smartlimiter_controller.go [hunk 5]
```diff
@@ -120,32 +129,130 @@ func (r *SmartLimiterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+
+	// register handler in framework, and trigger by event or ticker
+	pc.WatcherProducerConfig.NeedUpdateMetricHandler = r.handleWatcherEvent
+	pc.TickerProducerConfig.NeedUpdateMetricHandler = r.handleTickerEvent
+	// start producer
+	metric.NewProducer(pc)
+	log.Infof("producers starts")
+
+
+	// TODO 0124 如果开关没有打开那么
```
Comment (cywang1905) [inline controllers/smartlimiter_controller.go:161]: change to english better

File: controllers/smartlimiter_controller.go [hunk 6]
```diff
@@ -120,32 +129,130 @@ func (r *SmartLimiterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	if env.Config.Metric != nil {
```
Comment (cywang1905) [inline controllers/smartlimiter_controller.go:162]: if metric is nil, limiter will not watch metric. maybe it's better to write like this
```go
	if env.Config.Metric != nil {
		go r.WatchMetric()
	} else {
		log.Warningf("watching metric is not running")
	}

File: controllers/smartlimiter_controller.go [hunk 7]
```diff
@@ -120,32 +129,130 @@ func (r *SmartLimiterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		go r.WatchMetric()
+	}
+
 	return r
 }
+
+//func NewReconciler(cfg *v1alpha1.Limiter, mgr ctrl.Manager, env *bootstrap.Environment) *SmartLimiterReconciler {
```
Comment (cywang1905) [inline controllers/smartlimiter_controller.go:169]: remove useless code to keep clean

File: controllers/smartlimiter_controller.go [hunk 8]
```diff
@@ -120,32 +129,130 @@ func (r *SmartLimiterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+//		lastUpdatePolicyLock: &sync.RWMutex{},
+//		//globalRateLimitInfo: cmap.New(),
+//	}
+//	r.source.Start(env.Stop)
+//	r.WatchSource(env.Stop)
+//	return r
+//}
+
+
+// producerConfig 定义了
```
Comment (cywang1905) [inline controllers/smartlimiter_controller.go:200]: same as above

File: controllers/smartlimiter_controller.go [hunk 9]
```diff
@@ -120,32 +129,130 @@ func (r *SmartLimiterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		enablePrometheusSource = true
+		prometheusSourceConfig, err = newPrometheusSourceConfig(env)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, stderrors.New("wrong metric_source_type")
+	}
+
+	// TODO do not need to query prometheus 0124
```
Comment (cywang1905) [inline controllers/smartlimiter_controller.go:221]: it seems like these `TODO` have relations with each other, maybe it's a hint for future tasks, make it more clearly for other people?

**Review by cywang1905 — CHANGES_REQUESTED** [review]

File: controllers/process.go (hunk 1 above)
Comment (MouceL) [reply controllers/process.go:311]: it will add disableGlobalRatelimite swtich here

**Review by cywang1905 — APPROVED** [review]

File: .gitignore [hunk 10]
```diff
@@ -6,7 +6,7 @@
 *.so
 *.dylib
 bin
-
+Makefile
```
Comment (YonkaFang) [inline .gitignore:9]: it's strange to add `Makefile` to .gitignore

File: controllers/envoy_filters.go [hunk 11]
```diff
@@ -41,7 +42,7 @@ func (r *SmartLimiterReconciler) GenerateEnvoyConfigs(spec microservicev1alpha2.
 		} else {
 			log.Errorf("get svc %s:%s err: %+v", loc.Name, loc.Namespace, err.Error())
 		}
-		return setsEnvoyFilter, setsSmartLimitDescriptor, globalDescriptors,err
+		return setsEnvoyFilter, setsSmartLimitDescriptor, globalDescriptors, err
```
Comment (YonkaFang) [inline controllers/envoy_filters.go:45]: we'd better not mess the format-code with actual changes.

**Pull request merged**